File: src/main/java/com/timepath/hl2/io/demo/LZSS.java
package com.timepath.hl2.io.demo;

import org.apache.commons.io.EndianUtils;

public class LZSS {

    public static final String ID = "LZSS";
    private static final int LZSS_LOOKSHIFT = 4;

    public static byte[] inflate(byte[] input) throws LZSSException {
        int pInput = 8, pOutput = 0; // Pointers
        // Header
        String id = new String(input, 0, 4);
        int actualSize = EndianUtils.readSwappedInteger(input, 4);
        if (!ID.equals(id) || actualSize == 0) {
            throw new LZSSException("Unrecognized header");
        }
        // Payload
        byte[] output = new byte[actualSize];
        int totalBytes = 0, cmdByte = 0, getCmdByte = 0;
        for (; ; ) {
            if (getCmdByte == 0) {
                cmdByte = input[pInput++];
            }
            getCmdByte = (getCmdByte + 1) & 0x07;
            if ((cmdByte & 0x01) == 0x01) {
                int position = (input[pInput++] << LZSS_LOOKSHIFT) | (input[pInput] >> LZSS_LOOKSHIFT);
                int count = (input[pInput++] & 0x0F) + 1;
                if (count == 1) {
                    break;
                }
                int pSource = pOutput - position - 1;
                for (int i = 0; i < count; i++) {
                    output[pOutput++] = output[pSource++];
                }
                totalBytes += count;
            } else {
                output[pOutput++] = input[pInput++];
                totalBytes++;
            }
            cmdByte >>= 1;
        }
        // Verify
        if (totalBytes != actualSize) {
            throw new LZSSException("Unexpected failure: bytes read do not match expected size");
        }
        return output;
    }

    static class LZSSException extends Exception {

        public LZSSException(String message) {
            super(message);
        }
    }
}

File: src/main/kotlin/com/timepath/hl2/io/demo/LZSS.kt
package com.timepath.hl2.io.demo

import com.timepath.toUnsigned

public class LZSS {

    class LZSSException(message: String) : Exception(message)

    companion object {

        fun readSwappedInteger(data: ByteArray, offset: Int) = (0
                + (data[offset + 0].toUnsigned() shl 0)
                + (data[offset + 1].toUnsigned() shl 8)
                + (data[offset + 2].toUnsigned() shl 16)
                + (data[offset + 3].toUnsigned() shl 24)
                )

        public val ID: String = "LZSS"
        private val LZSS_LOOKSHIFT = 4

        throws(LZSSException::class)
        public fun inflate(input: ByteArray): ByteArray {
            // Pointers
            var pInput = 8
            var pOutput = 0
            // Header
            val id = String(input, 0, 4)
            val actualSize = readSwappedInteger(input, 4)
            if (ID != id || actualSize == 0) {
                throw LZSSException("Unrecognized header")
            }
            // Payload
            val output = ByteArray(actualSize)
            var totalBytes = 0
            var cmdByte = 0
            var getCmdByte = 0
            while (true) {
                if (getCmdByte == 0) {
                    cmdByte = input[pInput++].toInt()
                }
                getCmdByte = (getCmdByte + 1) and 7
                if ((cmdByte and 1) == 1) {
                    val position = (input[pInput++].toInt() shl LZSS_LOOKSHIFT) or (input[pInput].toInt() shr LZSS_LOOKSHIFT)
                    val count = (input[pInput++].toInt() and 15) + 1
                    if (count == 1) {
                        break
                    }
                    var pSource = pOutput - position - 1
                    for (i in 0..count - 1) {
                        output[pOutput++] = output[pSource++]
                    }
                    totalBytes += count
                } else {
                    output[pOutput++] = input[pInput++]
                    totalBytes++
                }
                cmdByte = cmdByte shr 1
            }
            // Verify
            if (totalBytes != actualSize) {
                throw LZSSException("Unexpected failure: bytes read do not match expected size")
            }
            return output
        }
    }
}
